### PR TITLE
Cluster small table/archetype into single Task in parallel iteration

### LINKED ISSUE #7303: Cluster small archetypes in parallel iteration
## What problem does this solve or what need does it fill?
`QueryParIter` establishes a batch size to chunk matching archetypes and tables into tasks to run on. Any archetype or table smaller than the batch size ends up in it's own task. This might help with work stealing, but it adds scheduling overhead.

## What solution would you like?
Collect batches smaller than the provided batch size into an `arrayvec` and dispatch it as a task when either the total count exceeds the batch size or the buffer is full.

## What alternative(s) have you considered?

Eating the performance cost and leaving it as is.

### PR DESCRIPTION
**Pull request by re0312**
# Objective

- Fix #7303
- bevy would spawn a lot of tasks in parallel iteration when it matchs a large storage and many small storage ,it significantly increase the overhead of schedule.

## Solution

- collect small storage into one task

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
Looks sensible, but I'd really like to see benchmarks on the impact of this change.

**Comment by re0312** [comment]
> Looks sensible, but I'd really like to see benchmarks on the impact of this change.

Added benchmark for fragment par-iter
image

have slight loss without fragment for par_iter , but a huge gain when fragments exceeds  available threads.

I also test `many_foxes` (which is not friendly for this PR's changes), it still showed a slight performance improvement. , the most significant performance gains came from the `check_visibility `
image

File: crates/bevy_ecs/src/query/state.rs [hunk 1]
```diff
@@ -1392,53 +1392,92 @@ impl<D: QueryData, F: QueryFilter> QueryState<D, F> {
             // SAFETY: We only access table data that has been registered in `self.archetype_component_access`.
             let tables = unsafe { &world.storages().tables };
             let archetypes = world.archetypes();
-            for storage_id in &self.matched_storage_ids {
-                if D::IS_DENSE && F::IS_DENSE {
-                    let table_id = storage_id.table_id;
-                    let table = &tables[table_id];
-                    if table.is_empty() {
-                        continue;
+            let mut batch_queue = vec![];
```
Comment (james7132) [inline crates/bevy_ecs/src/query/state.rs:1395]: I'm not the biggest fan of needing to allocate a growable vec repeatedly. Could you try using `arrayvec`, with a reasonable max queue size (128 seems more than reasonable).

**Review by james7132 — COMMENTED** [review]
I'm a bit concerned what this will do with hierarchical parallel queries (i.e. transform propagation), where the top level parallel query has only a few values.

Otherwise looks good! Just one bit about allocation.

File: crates/bevy_ecs/src/query/state.rs (hunk 1 above)
Comment (re0312) [reply crates/bevy_ecs/src/query/state.rs]: benchmark using arrayvec

image

File: crates/bevy_ecs/Cargo.toml [hunk 2]
```diff
@@ -30,6 +30,7 @@ rustc-hash = "1.1"
 serde = "1"
 thiserror = "1.0"
 nonmax = "0.5"
+arrayvec = "0.7.4"
```
Comment (james7132) [inline crates/bevy_ecs/Cargo.toml:33]: This should be optional and enabled with `multi-threaded`

**Review by james7132 — APPROVED** [review]

**Review by SkiFire13 — APPROVED** [review]

**Pull request merged**